### PR TITLE
feat: management dashboard v1

### LINKED ISSUE #131: Monitoring dashboard
Create a dashboard page showing relevant data related to :

- Consumers
- Producers
- Lag
- Throughput
- Searching data
- etc

### LINKED ISSUE #178: Dashboard - Consumers management
Dashboard with all the partitions assignment of each consumer, allowing to manage a consumer in all instances/machines.

It must have the following features:
- Pause the consumption
- Restart the consumption
- Resume the consumption 
- Reset the partition offset
- Rewind the partition offset
- Change the number of workers

### PR DESCRIPTION
**Pull request by dougolima**
# Description

Create a management dashboard using angular and bootstrap

Fixes #178

## How Has This Been Tested?

Unit test and visualizing data in the dashboard page

## Checklist

-   [x] My code follows the style guidelines of this project
-   [x] I have performed a self-review of my own code
-   [x] I have added tests to cover my changes
-   [ ] I have made corresponding changes to the documentation

### Disclaimer

By sending us your contributions, you are agreeing that your contribution is made subject to the terms of our [Contributor Ownership Statement](https://github.com/Farfetch/.github/blob/master/COS.md)

### CODE REVIEW

File: src/KafkaFlow.Admin/Extensions/ClusterConfigurationBuilderExtensions.cs [hunk 1]
```diff
@@ -58,5 +67,97 @@ public static class ClusterConfigurationBuilderExtensions
...
+                                async _ =>
+                                {
+                                    await resolver
+                                        .Resolve<IProducerAccessor>()
+                                        .GetProducer(producerName)
+                                        .BatchProduceAsync(
+                                            resolver
+                                                .Resolve<IConsumerAccessor>()
+                                                .All
+                                                .Where(c => !c.IsReadonly) // TODO GET ONLY CONSUMERS FROM THIS CLUSTER
```
Comment (dougolima) [inline src/KafkaFlow.Admin/Extensions/ClusterConfigurationBuilderExtensions.cs:120]: How to get the consumers by Cluster? 
Maybe defining a Name/Id in the cluster and exposing it on Consumer interface (ClusterId/Name), so we could filter the consumers here.

File: samples/KafkaFlow.Sample.Dashboard/KafkaFlow.Sample.Dashboard.csproj [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    </Exec>
+    <Error Condition="'$(ErrorCode)' != '0'" Text="Node.js is required to build and run this project. To continue, please install Node.js from https://nodejs.org/, and then restart your command prompt or IDE." />
+    <Message Importance="high" Text="Restoring dependencies using 'npm'. This may take several minutes..." />
+    <Exec WorkingDirectory="$(OutDir)$(SpaRoot)" Command="npm install" />
+  </Target>
+
+  <Target Name="PublishRunWebpack" AfterTargets="ComputeFilesToPublish">
+    <!-- As part of publishing, ensure the JS resources are freshly built in production mode -->
+    <Exec WorkingDirectory="$(OutDir)$(SpaRoot)" Command="npm run build -- --prod" />
+    <Exec WorkingDirectory="$(OutDir)$(SpaRoot)" Command="if not exist $(PublishDir)$(SpaRoot)dist mkdir $(PublishDir)$(SpaRoot)dist"/>
```
Comment (dougolima) [inline samples/KafkaFlow.Sample.Dashboard/KafkaFlow.Sample.Dashboard.csproj:31]: How will these windows commands work in a linux environments?

Comment (dougolima) [reply samples/KafkaFlow.Sample.Dashboard/KafkaFlow.Sample.Dashboard.csproj]: Solved, created commands based on operational system

File: src/KafkaFlow.Admin/Extensions/ClusterConfigurationBuilderExtensions.cs [hunk 3]
```diff
@@ -58,5 +67,97 @@ public static class ClusterConfigurationBuilderExtensions
...
+                                                .SelectMany(c => c.Assignment.Select(a =>
+                                                    new BatchProduceItem(
+                                                        topicName,
+                                                        Guid.NewGuid().ToString(),
```
Comment (filipeesch) [inline src/KafkaFlow.Admin/Extensions/ClusterConfigurationBuilderExtensions.cs:124]: In KF 2.0, you can remove the ToString(0 and use the GetBytes()

File: src/KafkaFlow.Admin/Extensions/ClusterConfigurationBuilderExtensions.cs [hunk 4]
```diff
@@ -58,5 +67,97 @@ public static class ClusterConfigurationBuilderExtensions
...
+                    consumer => consumer
+                        .Topic(topicName)
+                        .WithGroupId(groupId)
+                        .WithWorkersCount(1)
+                        .AsReadonly()
+                        .WithBufferSize(10)
+                        .WithAutoOffsetReset(AutoOffsetReset.Latest)
+                        .WithPartitionsAssignedHandler((resolver, partitions) =>
+                        {
+                            TelemetryScheduler.Set(
```
Comment (filipeesch) [inline src/KafkaFlow.Admin/Extensions/ClusterConfigurationBuilderExtensions.cs:109]: I think it would be better to move this logic to a separate class and use the dependence injection here too

File: src/KafkaFlow.Admin/Handlers/ConsumerMetricHandler.cs [hunk 5]
```diff
@@ -0,0 +1,19 @@
+namespace KafkaFlow.Admin.Handlers
+{
+    using System.Threading.Tasks;
+    using KafkaFlow.Admin.Messages;
+    using KafkaFlow.TypedHandler;
+
+    internal class ConsumerMetricHandler : IMessageHandler<ConsumerMetric>
+    {
+        private readonly ITelemetryCache cache;
```
Comment (filipeesch) [inline src/KafkaFlow.Admin/Handlers/ConsumerMetricHandler.cs:9]: Changing the ITelemetryCache interface name, we need to change the names of the fields too.

File: src/KafkaFlow.Admin/ITelemetryCache.cs [hunk 6]
```diff
@@ -0,0 +1,27 @@
+namespace KafkaFlow.Admin
+{
+    using System.Collections.Generic;
+    using KafkaFlow.Admin.Messages;
+
+    /// <summary>
+    /// Used to implement a cache provider to manage telemetry data
+    /// </summary>
+    public interface ITelemetryCache
```
Comment (filipeesch) [inline src/KafkaFlow.Admin/ITelemetryCache.cs:9]: I think we can change the name to ITelemetryStore(age?)

File: src/KafkaFlow.Admin/InMemoryTelemetryCache.cs [hunk 7]
```diff
@@ -0,0 +1,40 @@
+namespace KafkaFlow.Admin
+{
+    using System.Collections.Generic;
+    using KafkaFlow.Admin.Messages;
+    using Microsoft.Extensions.Caching.Memory;
+
+    /// <summary>
+    /// Provide cache operations related to telemetry data
+    /// </summary>
+    public class InMemoryTelemetryCache : ITelemetryCache
```
Comment (filipeesch) [inline src/KafkaFlow.Admin/InMemoryTelemetryCache.cs:10]: MemoryCacheTelemetryStore?

File: src/KafkaFlow.Admin/Messages/ConsumerMetric.cs [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Runtime.Serialization;
+
+    /// <summary>
+    /// A message that contains data related to consumers partition assigment
+    /// </summary>
+    [DataContract]
+    public class ConsumerMetric
```
Comment (filipeesch) [inline src/KafkaFlow.Admin/Messages/ConsumerMetric.cs:11]: ConsumerTelemetry maybe?

File: src/KafkaFlow/Consumers/ConsumerFlowManager.cs [hunk 9]
```diff
@@ -121,6 +121,14 @@ public void Resume(IReadOnlyCollection<TopicPartition> topicPartitions)
...
         }
 
+        public void UpdatePausedPartitions(IEnumerable<TopicPartition> partitionsRunning)
+        {
+            foreach (var p in partitionsRunning)
+            {
+                this.pausedPartitions.Remove(p);
+            }
+        }
+
```
Comment (filipeesch) [inline src/KafkaFlow/Consumers/ConsumerFlowManager.cs:131]: I think there are other approaches instead of using this method.

File: src/KafkaFlow/Consumers/IMessageConsumer.cs [hunk 10]
```diff
@@ -49,21 +59,21 @@ public interface IMessageConsumer
...
-        /// <summary>
-        /// Gets the current number of workers allocated of the consumer
-        /// </summary>
-        int WorkerCount { get; }
-
         /// <summary>
         /// Gets the current consumer flow status
         /// </summary>
-        ConsumerFlowStatus FlowStatus { get; }
+        ConsumerFlowStatus? FlowStatus { get; }
```
Comment (filipeesch) [inline src/KafkaFlow/Consumers/IMessageConsumer.cs:65]: Why the change to nullable?

File: src/KafkaFlow/Consumers/MessageConsumer.cs [hunk 11]
```diff
@@ -21,27 +21,37 @@ internal class MessageConsumer : IMessageConsumer
...
 
         public string MemberId => this.consumerManager.Consumer.MemberId;
 
         public string ClientInstanceName => this.consumerManager.Consumer.ClientInstanceName;
 
-        public int WorkerCount => this.consumerManager.Consumer.Configuration.WorkerCount;
+        public int WorkersCount => this.consumerManager.Consumer.Configuration.WorkersCount;
 
-        public ConsumerFlowStatus FlowStatus => this.consumerManager.Consumer.FlowManager.Status;
+        public IReadOnlyList<TopicPartition> PausedPartitions => this.consumerManager.Consumer.FlowManager?.PausedPartitions ?? new List<TopicPartition>();
```
Comment (filipeesch) [inline src/KafkaFlow/Consumers/MessageConsumer.cs:40]: I think we can use an IEnumerable here and return Enumerable.Empty() instead of new List()

File: src/KafkaFlow/Consumers/MessageConsumer.cs [hunk 12]
```diff
@@ -21,27 +21,37 @@ internal class MessageConsumer : IMessageConsumer
...
 
-        public IReadOnlyList<TopicPartition> PausedPartitions => this.consumerManager.Consumer.FlowManager.PausedPartitions;
+        public IReadOnlyList<TopicPartition> RunningPartitions => this.Assignment.Except(this.PausedPartitions).ToList();
```
Comment (filipeesch) [inline src/KafkaFlow/Consumers/MessageConsumer.cs:42]: I would use IEnumerable here too and remove the ToList()

File: src/KafkaFlow/KafkaBus.cs [hunk 13]
```diff
@@ -58,6 +58,8 @@ public async Task StartAsync(CancellationToken stopCancellationToken = default)
 
         public Task StopAsync()
         {
+            this.configuration.Clusters.ToList().ForEach(c => c.OnStopHandler(this.dependencyResolver));
```
Comment (filipeesch) [inline src/KafkaFlow/KafkaBus.cs:61]: I would use a foreach loop intead of ToList().ForEach()

File: src/KafkaFlow.Admin/TelemetryScheduler.cs [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+        public static void Unset(string key)
+        {
+            if (Timers.Value.TryGetValue(key, out var timer))
+            {
+                timer.Dispose();
+                Timers.Value.Remove(key);
+            }
+        }
+    }
+}
```
Comment (filipeesch) [inline src/KafkaFlow.Admin/TelemetryScheduler.cs:27]: Maybe we have to do some changes here

**Review by filipeesch — CHANGES_REQUESTED** [review]

File: src/KafkaFlow.Admin.Dashboard/AngularFiles/src/app/consumer.service.ts [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+
+  constructor(private http: HttpClient) {
+    this.headers = new HttpHeaders({'Content-Type': 'application/json; charset=utf-8'});
+  }
+
+  public get() {
+    return this.http.get(this.accessPointUrl + 'groups', {headers: this.headers});
+  }
+
+  public updateWorkersCount(groupId: any, consumerName: any, workersCount: number) {
```
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow.Admin.Dashboard/AngularFiles/src/app/consumer.service.ts:20]: Yeah, I agree... it was a forgotten thing.. I will try to remove all the `any`

File: samples/KafkaFlow.Sample.Dashboard/Startup.cs [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+                .UseEndpoints(endpoints => { endpoints.MapControllers(); })
+                .UseKafkaFlowDashboard(env);
+
+            lifetime.ApplicationStarted.Register(() =>
+            {
+                bus = app.ApplicationServices.CreateKafkaBus();
+                bus.StartAsync(lifetime.ApplicationStopped);
+            });
+            
+            lifetime.ApplicationStopping.Register(() => bus.StopAsync().GetAwaiter().GetResult());
```
[the remark this answers is not in the document]
Comment (dougolima) [reply samples/KafkaFlow.Sample.Dashboard/Startup.cs:55]: Right, done.

File: src/KafkaFlow.Abstractions/Configuration/IConsumerConfigurationBuilder.cs [hunk 17]
```diff
@@ -41,6 +41,12 @@ public interface IConsumerConfigurationBuilder
         /// <returns></returns>
         IConsumerConfigurationBuilder WithName(string name);
 
+        /// <summary>
+        /// Sets the consumer as readonly, that means this consumer can not be managed and it will not send telemetry data 
+        /// </summary>
+        /// <returns></returns>
+        IConsumerConfigurationBuilder AsReadonly();
```
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow.Abstractions/Configuration/IConsumerConfigurationBuilder.cs:48]: Another options:
- DisableManagement
- UnManageable
- WithoutManagementOperations

File: src/KafkaFlow.Admin.Dashboard/AngularFiles/src/app/consumer.service.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow.Admin.Dashboard/AngularFiles/src/app/consumer.service.ts:20]: done

File: src/KafkaFlow/Consumers/IMessageConsumer.cs (hunk 10 above)
Comment (dougolima) [reply src/KafkaFlow/Consumers/IMessageConsumer.cs:65]: Because when we are restarting the consumer, the `Consumer.InvalidateConsumer` method sets null into `FlowManager` property.
 
If during this time interval someone tries to retrieve information about this consumer (through the web api, for example) it would occur an exception.

To avoid this behaviour, the FlowManager now is nullable and the ConsumerResponseAdapter returns `ConsumerFlowStatus.NotRunning` status if consumer.FlowStatus.

Comment (dougolima) [reply src/KafkaFlow/Consumers/IMessageConsumer.cs:65]: Another option would be to put this logic directly into `MessageConsumer`, keeping the `ConsumerFlowStatus` as not nullable and returning `ConsumerFlowStatus.NotRunning` when `FlowManager` is null.

Actually, I think it would be better this approach.... what do you think?

File: src/KafkaFlow.Admin/Extensions/ClusterConfigurationBuilderExtensions.cs (hunk 1 above)
Comment (dougolima) [reply src/KafkaFlow.Admin/Extensions/ClusterConfigurationBuilderExtensions.cs]: done

Comment (filipeesch) [reply src/KafkaFlow.Admin/Extensions/ClusterConfigurationBuilderExtensions.cs]: Let's create the OnStarted() on cluster and put this logic there

File: src/KafkaFlow.Abstractions/Configuration/IConsumerConfigurationBuilder.cs (hunk 17 above)
[the remark this answers is not in the document]
Comment (filipeesch) [reply src/KafkaFlow.Abstractions/Configuration/IConsumerConfigurationBuilder.cs:48]: I like DiableManagement :)

File: src/KafkaFlow.Admin.Dashboard/AngularFiles/src/app/home/home.component.ts [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+              "Paused" :
+              "Not Running";
+        c.partitionAssignments.forEach(function (pa: any) {
+          pa.isLost = !self.isActive(pa.lastUpdate);
+        })
+      })
+    });
+
+    return groups;
+  }
```
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow.Admin.Dashboard/AngularFiles/src/app/home/home.component.ts:32]: I will move it to consumer component

File: src/KafkaFlow/Consumers/IMessageConsumer.cs (hunk 10 above)
Comment (filipeesch) [reply src/KafkaFlow/Consumers/IMessageConsumer.cs:70]: We can move the ConsumerFlowStatus to Consumer class, calculate the NotRunning status based on FlowStatus is null or partitions assignment is empty or null.

Rename ConsumerFlowStatus to ConsumerStatus too

File: src/KafkaFlow.Admin/Extensions/ClusterConfigurationBuilderExtensions.cs (hunk 1 above)
Comment (dougolima) [reply src/KafkaFlow.Admin/Extensions/ClusterConfigurationBuilderExtensions.cs]: done

File: src/KafkaFlow.Abstractions/Configuration/IConsumerConfigurationBuilder.cs (hunk 17 above)
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow.Abstractions/Configuration/IConsumerConfigurationBuilder.cs:48]: Changed it.

File: src/KafkaFlow.Admin/TelemetryScheduler.cs (hunk 14 above)
Comment (dougolima) [reply src/KafkaFlow.Admin/TelemetryScheduler.cs:27]: Changed it, please take a look.

File: src/KafkaFlow.Admin.Dashboard/KafkaFlow.Admin.Dashboard.csproj [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+    <TypeScriptToolsVersion>Latest</TypeScriptToolsVersion>
+    <IsPackable>true</IsPackable>
+    <SpaRoot>AngularFiles\</SpaRoot>
+    <DefaultItemExcludes>$(DefaultItemExcludes);$(SpaRoot)node_modules\**;$(SpaRoot)dist\**</DefaultItemExcludes>
+    <OutputType>Library</OutputType>    
+  </PropertyGroup>
+
+  <ItemGroup>
+    <Content Remove="$(SpaRoot)**" />
+    <Content Include="$(SpaRoot)**" Exclude="$(SpaRoot)node_modules\**;$(SpaRoot)dist\**">
```
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow.Admin.Dashboard/KafkaFlow.Admin.Dashboard.csproj:15]: This will be fixed on [this](https://github.com/Farfetch/kafka-flow/pull/174#discussion_r647419741) issue.

File: src/KafkaFlow.Admin.WebApi/Adapters/ConsumerResponseAdapter.cs [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+
+        private static IEnumerable<PartitionAssignment> GetLocalInfo(IMessageConsumer consumer)
+        {
+            return consumer.PausedPartitions
+                .GroupBy(c => c.Topic)
+                .Select(c => new PartitionAssignment()
+                {
+                    Topic = c.Key,
+                    HostName = Environment.MachineName,
+                    PausedPartitions = c.Select(x => x.Partition.Value).ToList(),
```
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow.Admin.WebApi/Adapters/ConsumerResponseAdapter.cs:43]: done

File: src/KafkaFlow.Admin.WebApi/Adapters/ConsumerResponseAdapter.cs [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+                    LastUpdate = DateTime.Now,
+                })
+                .Union(consumer.RunningPartitions
+                    .GroupBy(c => c.Topic)
+                    .Select(c => new PartitionAssignment()
+                    {
+                        Topic = c.Key,
+                        HostName = Environment.MachineName,
+                        RunningPartitions = c.Select(x => x.Partition.Value).ToList(),
```
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow.Admin.WebApi/Adapters/ConsumerResponseAdapter.cs:52]: done

File: src/KafkaFlow.Admin.WebApi/Contracts/PartitionAssignment.cs [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+    {
+        /// <summary>
+        /// Gets or sets the topic name
+        /// </summary>
+        public string Topic { get; set; }
+
+        /// <summary>
+        /// Gets or sets the host name
+        /// </summary>
+        public string HostName { get; set; }
```
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow.Admin.WebApi/Contracts/PartitionAssignment.cs:19]: Changed to InstanceName.

File: src/KafkaFlow.Admin/Extensions/ClusterConfigurationBuilderExtensions.cs (hunk 3 above)
Comment (dougolima) [reply src/KafkaFlow.Admin/Extensions/ClusterConfigurationBuilderExtensions.cs]: done

File: src/KafkaFlow.Admin/Extensions/ClusterConfigurationBuilderExtensions.cs (hunk 4 above)
Comment (dougolima) [reply src/KafkaFlow.Admin/Extensions/ClusterConfigurationBuilderExtensions.cs]: Changed, please take a look along with [this](https://github.com/Farfetch/kafka-flow/pull/174#discussion_r647497190) issue.

File: src/KafkaFlow.Admin/ITelemetryCache.cs (hunk 6 above)
Comment (dougolima) [reply src/KafkaFlow.Admin/ITelemetryCache.cs]: done

File: samples/KafkaFlow.Sample.Dashboard/KafkaFlow.Sample.Dashboard.csproj [hunk 23]
```diff
@@ -0,0 +1,39 @@
...
+  <Target Name="DebugNpmInstall" BeforeTargets="Build">
+    <Exec Command="node --version" ContinueOnError="true">
+      <Output TaskParameter="ExitCode" PropertyName="ErrorCode" />
+    </Exec>
+    <Error Condition="'$(ErrorCode)' != '0'" Text="Node.js is required to build and run this project. To continue, please install Node.js from https://nodejs.org/, and then restart your command prompt or IDE." />
+    <Message Importance="high" Text="Restoring dependencies using 'npm'. This may take several minutes..." />
+    <Exec WorkingDirectory="$(OutDir)$(SpaRoot)" Command="npm install" />
+  </Target>
+
+  <Target Name="PublishRunWebpack" AfterTargets="ComputeFilesToPublish">
```
[the remark this answers is not in the document]
Comment (dougolima) [reply samples/KafkaFlow.Sample.Dashboard/KafkaFlow.Sample.Dashboard.csproj]: done

File: src/KafkaFlow.Admin/Messages/ConsumerMetric.cs (hunk 8 above)
Comment (dougolima) [reply src/KafkaFlow.Admin/Messages/ConsumerMetric.cs:11]: IDK, in my opinion, ConsumerMetric (the metrics of consumers) fits better... but if you want we can change it.

File: src/KafkaFlow.Admin.Dashboard/AngularFiles/src/index.html [hunk 24]
```diff
@@ -0,0 +1,16 @@
+<!doctype html>
+<html lang="en">
+<head>
+  <meta charset="utf-8">
+  <title>KafkaFlow - Dashboard</title>
+  <base href="/">
+  <meta name="viewport" content="width=device-width, initial-scale=1">
+  <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@4.6.0/dist/css/bootstrap.min.css" integrity="sha384-B0vP5xmATw1+K9KRQjQERJvTumQW0nPEzvF6L/Z6nronJ3oUOFUFpCjEUQouq2+l" crossorigin="anonymous">
+  <script src="https://code.jquery.com/jquery-3.5.1.slim.min.js" integrity="sha384-DfXdz2htPH0lsSSs5nCTpuj/zy4C+OGpamoFVy38MVBnE+IbbVYUew+OrCXaRkfj" crossorigin="anonymous"></script>
```
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow.Admin.Dashboard/AngularFiles/src/index.html]: It was necessary at some point but it is not anymore. Deleted it.

File: src/KafkaFlow.Admin.WebApi/Contracts/ConsumerResponse.cs [hunk 25]
```diff
@@ -0,0 +1,63 @@
...
+       ///     instances in a given application which allows
+       ///     log messages to be associated with the
+       ///     corresponding instance.
+       /// </remarks>
+       public string ClientInstanceName { get; set; }
+
+       /// <summary>
+       /// Gets or sets the current consumer flow status
+       /// </summary>
+       public ConsumerFlowStatus FlowStatus { get; set; }
```
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow.Admin.WebApi/Contracts/ConsumerResponse.cs:61]: done

File: src/KafkaFlow.Admin/Messages/ConsumerMetric.cs (hunk 8 above)
Comment (dougolima) [reply src/KafkaFlow.Admin/Messages/ConsumerMetric.cs:12]: done

File: src/KafkaFlow.Admin.WebApi/Contracts/ConsumerResponse.cs [hunk 26]
```diff
@@ -0,0 +1,63 @@
...
+       ///     instances in a given application which allows
+       ///     log messages to be associated with the
+       ///     corresponding instance.
+       /// </remarks>
+       public string ClientInstanceName { get; set; }
+
+       /// <summary>
+       /// Gets or sets the current consumer flow status
+       /// </summary>
+       public string FlowStatus { get; set; }
```
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow.Admin.WebApi/Contracts/ConsumerResponse.cs:55]: done

File: src/KafkaFlow.Admin.WebApi/Controllers/TelemetryController.cs [hunk 27]
```diff
@@ -0,0 +1,52 @@
...
+        private readonly ITelemetryStorage storage;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="TelemetryController"/> class.
+        /// </summary>
+        /// <param name="consumers">The accessor class that provides access to the consumers</param>
+        /// <param name="storage">The cache interface to get metric data</param>
+        public TelemetryController(IConsumerAccessor consumers, ITelemetryStorage storage)
+        {
+            this.consumers = consumers;
```
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow.Admin.WebApi/Controllers/TelemetryController.cs]: done

File: src/KafkaFlow.Admin/ITelemetryScheduler.cs [hunk 28]
```diff
@@ -0,0 +1,9 @@
+namespace KafkaFlow.Admin
+{
+    internal interface ITelemetryScheduler
+    {
+        void Start(string key, string topicName);
```
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow.Admin/ITelemetryScheduler.cs]: Changed to `telemetryId`

File: src/KafkaFlow.Admin.WebApi/Controllers/ConsumersController.cs [hunk 29]
```diff
@@ -35,10 +39,14 @@ public ConsumersController(IConsumerAccessor consumers, IAdminProducer adminProd
         /// <param name="groupId">Identifier of the group</param>
         /// <returns>A list of consumers</returns>
         [HttpGet]
-        [ProducesResponseType(typeof(IEnumerable<IMessageConsumer>), 200)]
+        [ProducesResponseType(typeof(IEnumerable<ConsumerResponse>), 200)]
```
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow.Admin.WebApi/Controllers/ConsumersController.cs:39]: This endpoint returns a list of consumers with the group id provided, that is why is a collection of ConsumerResponse

File: src/KafkaFlow/Consumers/ConsumerFlowManager.cs [hunk 30]
```diff
@@ -24,7 +24,7 @@ internal class ConsumerFlowManager : IConsumerFlowManager
             this.logHandler = logHandler;
         }
 
-        public IReadOnlyList<TopicPartition> PausedPartitions => this.pausedPartitions.AsReadOnly();
+        public IEnumerable<TopicPartition> PausedPartitions => this.pausedPartitions.AsReadOnly();
```
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow/Consumers/ConsumerFlowManager.cs:27]: Changed to use `AsEnumerable()`

File: src/KafkaFlow.Admin.WebApi/Controllers/ConsumersController.cs (hunk 29 above)
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow.Admin.WebApi/Controllers/ConsumersController.cs:39]: done

**Review by filipeesch — CHANGES_REQUESTED** [review]

File: src/KafkaFlow/Consumers/Consumer.cs [hunk 31]
```diff
@@ -95,7 +119,7 @@ internal class Consumer : IConsumer
                 try
                 {
                     this.EnsureConsumer();
-
+                    await this.flowManager.SemaphoreWait();
```
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow/Consumers/Consumer.cs:122]: sure, any suggestion? Maybe just Wait?

File: src/KafkaFlow.Admin.Dashboard/ClientApp/src/app/group-by.pipe.ts [hunk 32]
```diff
@@ -0,0 +1,26 @@
...
+        previous[current[property]].push(current);
+      }
+
+      return previous;
+    }, {});
+
+    // this will return an array of objects, each object containing a group of objects
+    return Object.keys(groupedCollection).map(key => ({ key, value: groupedCollection[key] }));
+  }
+}
```
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow.Admin.Dashboard/ClientApp/src/app/group-by.pipe.ts:26]: yes, we are using it on consumer.component.html to group all partition assignment data by topic

File: src/KafkaFlow/Consumers/Consumer.cs [hunk 33]
```diff
@@ -137,32 +165,33 @@ private void EnsureConsumer()
...
         {
-            this.FlowManager?.Dispose();
-            this.FlowManager = null;
-
             this.consumer?.Close();
             this.consumer = null;
         }
     }
-}
+}
```
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow/Consumers/Consumer.cs:197]: done

File: src/KafkaFlow/Configuration/IConsumerConfiguration.cs [hunk 34]
```diff
@@ -30,9 +30,19 @@ public interface IConsumerConfiguration
         string ConsumerName { get; }
 
         /// <summary>
-        /// Gets or sets the current number of workers
+        /// Gets the cluster name
         /// </summary>
-        int WorkerCount { get; set; }
+        string ClusterName { get; }
```
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow/Configuration/IConsumerConfiguration.cs:35]: For now, we don't need any other cluster information except the name.

I think if someday we need it, so we can adapt the ConsumerConfiguration to have the entire ClusterConfiguration.

File: src/KafkaFlow/Consumers/Consumer.cs (hunk 33 above)
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow/Consumers/Consumer.cs]: done

File: src/KafkaFlow.Admin.WebApi/Controllers/ConsumersController.cs [hunk 35]
```diff
@@ -34,22 +35,29 @@ public ConsumersController(IConsumerAccessor consumers, IAdminProducer adminProd
         /// </summary>
         /// <param name="groupId">Identifier of the group</param>
         /// <returns>A list of consumers</returns>
-        [HttpGet]
-        [ProducesResponseType(typeof(IEnumerable<IMessageConsumer>), 200)]
+        [HttpGet(Name="GetConsumersByGroupId")]
```
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow.Admin.WebApi/Controllers/ConsumersController.cs:38]: Unfortunately not 😢

File: src/KafkaFlow.Admin.WebApi/Controllers/TelemetryController.cs [hunk 36]
```diff
@@ -0,0 +1,38 @@
...
+        public TelemetryController(ITelemetryStorage storage)
+        {
+            this.storage = storage;
+        }
+
+        /// <summary>
+        /// Get telemetry data from all the consumer groups
+        /// </summary>
+        /// <returns>A telemetry response</returns>
+        [HttpGet(Name="GetTelemetry")]
```
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow.Admin.WebApi/Controllers/TelemetryController.cs:29]: same answer 😢

**Review by filipeesch — APPROVED** [review]

File: src/KafkaFlow/Consumers/Consumer.cs (hunk 31 above)
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow/Consumers/Consumer.cs:122]: Changed to BlockHeartbeat and ReleaseHeartbeat
